pkg/cli/plugins_test.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cli
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"sigs.k8s.io/kubebuilder/pkg/plugin"
+)
+
+var _ = Describe("resolvePluginsByKey", func() {
+
+	var (
+		plugins = makePluginsForKeys(
+			"foo.example.com/v1.0",
+			"bar.example.com/v1.0",
+			"baz.example.com/v1.0",
+			"foo.kubebuilder.io/v1.0",
+			"foo.kubebuilder.io/v2.0",
+			"bar.kubebuilder.io/v1.0",
+			"bar.kubebuilder.io/v2.0",
+		)
+		resolvedPlugins []plugin.Base
+		err             error
+	)
+
+	It("should resolve keys correctly", func() {
+		By("resolving foo.example.com/v1.0")
+		resolvedPlugins, err = resolvePluginsByKey(plugins, "foo.example.com/v1.0")
+		Expect(err).NotTo(HaveOccurred())
+		Expect(makePluginKeySlice(resolvedPlugins...)).To(Equal([]string{"foo.example.com/v1.0"}))
+
+		By("resolving foo.example.com")
+		resolvedPlugins, err = resolvePluginsByKey(plugins, "foo.example.com")
+		Expect(err).NotTo(HaveOccurred())
+		Expect(makePluginKeySlice(resolvedPlugins...)).To(Equal([]string{"foo.example.com/v1.0"}))
+
+		By("resolving baz")
+		resolvedPlugins, err = resolvePluginsByKey(plugins, "baz")
+		Expect(err).NotTo(HaveOccurred())
+		Expect(makePluginKeySlice(resolvedPlugins...)).To(Equal([]string{"baz.example.com/v1.0"}))
+
+		By("resolving foo/v2.0")
+		resolvedPlugins, err = resolvePluginsByKey(plugins, "foo/v2.0")
+		Expect(err).NotTo(HaveOccurred())
+		Expect(makePluginKeySlice(resolvedPlugins...)).To(Equal([]string{"foo.kubebuilder.io/v2.0"}))
+	})
+
+	It("should return an error", func() {
+		By("resolving foo.kubebuilder.io")
+		_, err = resolvePluginsByKey(plugins, "foo.kubebuilder.io")
+		Expect(err).To(MatchError(errAmbiguousPlugin{
+			key: "foo.kubebuilder.io",
+			msg: `possible keys: ["foo.kubebuilder.io/v1.0" "foo.kubebuilder.io/v2.0"]`,
+		}))
+
+		By("resolving foo/v1.0")
+		_, err = resolvePluginsByKey(plugins, "foo/v1.0")
+		Expect(err).To(MatchError(errAmbiguousPlugin{
+			key: "foo/v1.0",
+			msg: `possible names: ["foo.example.com" "foo.kubebuilder.io"]`,
+		}))
+
+		By("resolving foo")
+		_, err = resolvePluginsByKey(plugins, "foo")
+		Expect(err).To(MatchError(errAmbiguousPlugin{
+			key: "foo",
+			msg: `possible keys: ["foo.example.com/v1.0" "foo.kubebuilder.io/v1.0" "foo.kubebuilder.io/v2.0"]`,
```

Comment:
IMO this scenario make is better covered with the currently duplicated issue for plugins. 
I mean the plugin/v1.0.0 is duplicated in my POV is a more intuitive error for it.

Reply:
The plugins known to the `CLI` are not duplicated, i.e. there is no duplicate key in `plugins` here. The user-defined key `foo` is ambiguous because it can match multiple known plugins.